### ISSUE TITLE
__iter__ class method from metaclass not recognized

### ISSUE THREAD

**Issue by nwalters512** [microsoft/pyright#4685]
**Describe the bug**

If an `__iter__(cls)` function is declared on a class that's then used as a metaclass, Pyright seemingly doesn't know about that `__iter__(cls)` function.

**To Reproduce**

Consider the following code:

```python
from typing import Iterator


class StyleMeta(type):
    def __iter__(cls) -> Iterator[str]:
        yield "a"
        yield "b"
        yield "c"


class Style(metaclass=StyleMeta):
    pass


x: type[Style] = Style
print(list(x))
```

At runtime, this code works as expected:

```
$ python3 code.py
['a', 'b', 'c']
```

However, Pyright reports an error:

```
error: Argument of type "Type[Style]" cannot be assigned to parameter "__iterable" of type "_GetItemIterable[_T@iter]" in function "iter"
    "__getitem__" is not present (reportGeneralTypeIssues)
```

**Expected behavior**

I would expect Pyright to accurately reflect what happens at runtime, which is that `Style` is indeed an iterable.

**VS Code extension or command-line**

I'm using the CLI version of Pyright.

```
$ yarn pyright --version
pyright 1.1.295
```

**Additional context**

I'm not suggesting that making a class itself an iterable is necessarily a good idea, but there is real-world code that does this. See https://github.com/python/typeshed/pull/9803 for some related discussion.

- Actual code: https://github.com/pygments/pygments/blob/8f417de6a9f3dff6d86a02c9faa365d6828a1046/pygments/style.py#L162
- Type stubs for that code: https://github.com/python/typeshed/blob/4ca5ee98df5654d0db7f5b24cd2bd3b3fe54f313/stubs/Pygments/pygments/style.pyi#L26

**Comment by erictraut** [comment]
Pyright's behavior is correct here. The `Iterable` protocol defines `__iter__` as follows:

```python
class Iterable(Protocol[_T_co]):
    @abstractmethod
    def __iter__(self) -> Iterator[_T_co]: ...
```

This means that the first parameter can be specified as either positional or by keyword, so it must be named `self`. It cannot be named `cls` if you want the Iterable to be compatible with this protocol.

If you think that `Iterable` shouldn't accept a keyword argument, you could file a bug with the maintainers of typeshed to redefine `Iterable` as follows:

```python
class Iterable(Protocol[_T_co]):
    @abstractmethod
    def __iter__(self, /) -> Iterator[_T_co]: ...
```

Note the addition of `/` to indicate that `self` is a positional-only parameter.

**Comment by AlexWaygood** [comment]
I think this is an unfortunate decision.

I've never considered the name of the first parameter of protocol methods to be part of the declared interface, and I think it will lead to pretty suboptimal results if this is pyright's behaviour. I really doubt that it was the intention of the original authors of the `Iterable` protocol to exclude `__iter__` methods on metaclasses from being considered compatible with the `Iterable` interface. (Naming the first method `cls` for instance methods on metaclasses is very common, and arguably more idiomatic than naming the first method `self`, since instances of metaclasses are, after all, classes.)

Moreover, the situations in which you could possibly call an instance method with a keyword argument in Python are exceedingly rare: you would have to do something very tortuous, like this:

```python
class Foo:
    def __iter__(self):
        yield from range(10)

f = Foo()
Foo.__iter__(self=f)
```

I've never seen that done, and I can't think of any use cases for doing so. I think it would be far better if pyright treated the first parameter of instance or class methods as being implicitly positional-only. I always assumed that this was already the behaviour that most type checkers had.

If you decide to keep to this decision, I will have to prepare a typeshed PR to change all protocols in typeshed so that the first parameter of every instance method in every protocol is explicitly declared as positional-only. This will be quite tedious and, in my opinion, quite unnecessary.

**Comment by erictraut** [comment]
@AlexWaygood, I don't see anything in [PEP 570](https://peps.python.org/pep-0570/) or [PEP 544](https://peps.python.org/pep-0544/) that would implicitly treat the first parameter in a method within a protocol as being position-only.

Are you confident that no caller ever relies on the name of the first parameter?

This might be a good discussion to move to the python/typing discussion forum. If there's consensus among the typing community, I'm willing to change pyright's behavior here.

I wouldn't recommend changing all of typeshed. Instead, I'd recommend that the OP change his code to use `self` rather than `cls` in this case. This is what typeshed does for `EnumMeta.__iter__`, for example.

**Comment by AlexWaygood** [comment]
> Are you confident that no caller ever relies on the name of the first parameter?

I'm pretty confident in this, yes.

> Instead, I'd recommend that the OP change his code to use self rather than cls in this case.

I know from the discussion on the typeshed PR https://github.com/python/typeshed/pull/9803 that this is impossible in this case: the metaclass with `__iter__` is in a third-party library that is outside of OP's control.

> I wouldn't recommend changing all of typeshed. Instead, I'd recommend that the OP change his code to use self rather than cls in this case. This is what typeshed does for EnumMeta.__iter__, for example.

In general, typeshed strongly tries to follow the runtime implementation wherever possible. Doing so greatly increases the accuracy of our stubs, as it allows tools such as mypy's stubtest script to easily compare the stubs with the runtime and check for inconsistencies. I consider it a bug here that the first parameter of `EnumMeta.__iter__` is called `self` in typeshed, as this is [not what the parameter is called at runtime](https://github.com/python/cpython/blob/e5e1c1fabd8b5626f9193e6c61b9d7ceb7fb2f95/Lib/enum.py#L782-L786).

**Comment by erictraut** [comment]
I will start a separate thread on python/typing later today to discuss this in more detail. I have a long series of follow-up questions that need to be answered before taking any action on this.

**Comment by AlexWaygood** [comment]
Sounds good. Thanks for your consideration!

**Comment by JelleZijlstra** [comment]
While the larger discussion Eric started on python/typing is important, wouldn't it be possible to solve the immediate problem here by marking the `self` parameter as positional-only on typing.Iterable in typeshed?

**Comment by AlexWaygood** [comment]
> While the larger discussion Eric started on python/typing is important, wouldn't it be possible to solve the immediate problem here by marking the `self` parameter as positional-only on typing.Iterable in typeshed?

Well, I don't see any reason that `typing.Iterable` should be seen as special here. Why not also change `typing.Sized`, `os.PathLike`, and all the other many protocols we have in typeshed? The same logic would seem to apply to all protocols, ever. This is why I wrote above:

> If you decide to keep to this decision, I will have to prepare a typeshed PR to change all protocols in typeshed so that the first parameter of every instance method in every protocol is explicitly declared as positional-only. This will be quite tedious and, in my opinion, quite unnecessary.

**Comment by erictraut** [comment]
Based on the discussion in python/typing, I've changed pyright's logic to assume that `self` and `cls` are positional-only parameters for purposes of protocol matching. I didn't go further and enforce positional-only semantics generally for instance methods because it doesn't match the runtime behavior.

**Comment by AlexWaygood** [comment]
Sounds good to me. Thanks again for being willing to reconsider here.

**Comment by erictraut** [comment]
This is included in pyright 1.1.296, which I just published. It will also be included in a future release of pylance.
